### PR TITLE
quic - fixed RETRY

### PR DESCRIPTION
**Pull request by nbridge-jump**
Fixes to RETRY handling
Also, robustness validity checks added

### CODE REVIEW

File: src/waltz/quic/fd_quic.c [hunk 1]
```diff
@@ -1443,19 +1436,33 @@ fd_quic_handle_v1_initial( fd_quic_t *               quic,
...
 
         /* This is the initial packet before retry. */
-        if (initial->token_len == 0)
-        {
+        if( initial->token_len == 0 ) {
+          /* FIXME we pass the "tp" here to allow the function to add the retry_source_connection_ip
+           * transport parameter to be added.
+           * This is incorrect, since:
+           *   1. "tp" is transient, and will likely be modified before it is encoded into a packet
+           *   2. RETRY is supposed to be stateless
```
Comment (lidatong) [inline src/waltz/quic/fd_quic.c:1451]: ah got it thanks

File: src/waltz/quic/fd_quic.c [hunk 2]
```diff
@@ -1467,23 +1474,67 @@ fd_quic_handle_v1_initial( fd_quic_t *               quic,
...
+		  original_destination_connection_id to S1 (note that this value is chosen by the client) and
+		  initial_source_connection_id to S3. In this case, the server does not include a
+                  retry_source_connection_id transport parameter.
+
+              When the handshake includes a Retry (Figure 8), the server sets
+                  original_destination_connection_id to S1, retry_source_connection_id
+                  to S2, and initial_source_connection_id to S3.  */
+        tp->original_destination_connection_id_present = 1;
+        tp->original_destination_connection_id_len     = orig_dst_conn_id.sz;
+        memcpy( tp->original_destination_connection_id,
```
Comment (lidatong) [inline src/waltz/quic/fd_quic.c:1528]:
```suggestion
        fd_memcpy( tp->original_destination_connection_id,
```

File: src/waltz/quic/fd_quic.c [hunk 3]
```diff
@@ -1467,23 +1474,67 @@ fd_quic_handle_v1_initial( fd_quic_t *               quic,
...
+                orig_dst_conn_id.conn_id,
+                orig_dst_conn_id.sz );
+
+        tp->retry_source_connection_id_present = 1;
+        tp->retry_source_connection_id_len     = retry_src_conn_id->sz;
+        memcpy( tp->retry_source_connection_id,
```
Comment (lidatong) [inline src/waltz/quic/fd_quic.c:1534]:
```suggestion
        fd_memcpy( tp->retry_source_connection_id,
```

File: src/waltz/quic/fd_quic.c [hunk 4]
```diff
@@ -1529,6 +1580,8 @@ fd_quic_handle_v1_initial( fd_quic_t *               quic,
 
       memcpy( &conn->orig_dst_conn_id, &orig_dst_conn_id, sizeof( orig_dst_conn_id ) );
 
+      /* keep retry_src_conn_id */
+      memcpy( &conn->retry_src_conn_id, retry_src_conn_id, sizeof( retry_src_conn_id ) );
```
Comment (lidatong) [inline src/waltz/quic/fd_quic.c:1584]:
```suggestion
      fd_memcpy( &conn->retry_src_conn_id, retry_src_conn_id, sizeof( retry_src_conn_id ) );
```

**Review by lidatong — APPROVED** [review]

File: src/waltz/quic/fd_quic.c (hunk 1 above)
Comment (nbridge-jump) [reply src/waltz/quic/fd_quic.c:1451]: That comment needs fixing. I believe the "tp" usage may have been correct. Either way "FIXME" isn't needed anymore

**Comment by nbridge-jump** [comment]
Think I'll just make a note of these and merge this.

**Pull request merged**